### ISSUE TITLE
Mixup Errors

### ISSUE THREAD

**Issue by jaideep11061982** [fastai/fastai#3183]
Be sure you've searched [the forums](https://forums.fast.ai) for the error message you received. Also, unless you're an experienced fastai developer, first ask on the forums to see if someone else has seen a similar issue already and knows how to solve it. **Only file a bug report here when you're quite confident it's not an issue with your local setup.**

**Please see [this model example](https://github.com/fastai/fastai2/issues/487) of how to fill out an issue correctly.  Please try to emulate that example as appropriate when opening an issue.**

Please confirm you have the latest versions of fastai, fastcore, and nbdev prior to reporting a bug (delete one): YES / NO

**Describe the bug**

Mix up/Cutmix always throws an error if we use loss function other than Cross entropy..

```
class CutMix(MixUp):
    def __init__(self, alpha=1.): 
        self.distrib = Beta(tensor(alpha), tensor(alpha))
    def before_batch(self):
        lam = self.distrib.sample().squeeze().to(self.x.device)
        shuffle = torch.randperm(self.y.size(0)).to(self.x.device)
        #print(self.yb)
        
        self.yb1 = tuple(L(self.yb).itemgot(shuffle))
        nx_dims = len(self.x.size())
        bs, c, h, w = self.x.shape
        rx, ry = w*self.distrib.sample(), h*self.distrib.sample()
        rw, rh = w*(1-lam).sqrt(), h*(1-lam).sqrt()
        x1 = (rx-rw/2).clamp(min=0).round().to(int)
        x2 = (rx+rw/2).clamp(max=w).round().to(int)
        y1 = (ry-rh/2).clamp(min=0).round().to(int)
        y2 = (ry+rh/2).clamp(max=h).round().to(int)
        self.learn.xb[0][:,:,y1:y2,x1:x2] = self.learn.xb[0][shuffle,:,y1:y2,x1:x2]
        self.lam = 1- float(x2-x1)*(y2-y1)/(h*w)
        
        if not self.stack_y:
            ny_dims = len(self.y.size())
            self.learn.yb = tuple(L(self.yb1,self.yb).map_zip(torch.lerp,weight=unsqueeze(self.lam, n=ny_dims-1)))
```
`expected dtype long int for `weights` but got dtype float`
**To Reproduce**
1. 
```
class FocalCosineLoss(nn.Module):
    def __init__(self, alpha=1, gamma=2, xent=.1):
        super(FocalCosineLoss, self).__init__()
        self.alpha = alpha
        self.gamma = gamma

        self.xent = nn.Parameter(torch.tensor([xent])).to(device)

        self.y = torch.Tensor([1]).cuda()
        store_attr()

    def forward(self, input, target, reduction="mean"):
        cosine_loss = F.cosine_embedding_loss(input,
                                              F.one_hot(target, num_classes=input.size(-1)), self.y, reduction=reduction)

        cent_loss = F.cross_entropy(F.normalize(input), target, reduce=False)
        pt = torch.exp(-cent_loss)
        focal_loss = self.alpha * (1-pt)**self.gamma * cent_loss

        if reduction == "mean":
            focal_loss = torch.mean(focal_loss)

        return cosine_loss + self.xent * focal_loss
```
with Learner(dls, nn.Linear(3,4), loss_func=FocalCosineLoss(), cbs=mixup) as learn:
    
    learn.epoch,learn.training = 0,True
    learn.dl = dls.train
    b = dls.one_batch()
    learn._split(b)
    learn('before_batch')

_,axs = plt.subplots(3,3, figsize=(9,9))
dls.show_batch(b=(mixup.xb[0],mixup.y), ctxs=axs.flatten())
**Expected behavior**

**Error with full stack trace**
```
----> 7     learn('before_batch')
      8 
      9 _,axs = plt.subplots(3,3, figsize=(9,9))

/opt/conda/lib/python3.7/site-packages/fastai/learner.py in __call__(self, event_name)
    131     def ordered_cbs(self, event): return [cb for cb in sort_by_run(self.cbs) if hasattr(cb, event)]
    132 
--> 133     def __call__(self, event_name): L(event_name).map(self._call_one)
    134 
    135     def _call_one(self, event_name):

/opt/conda/lib/python3.7/site-packages/fastcore/foundation.py in map(self, f, gen, *args, **kwargs)
    224     def range(cls, a, b=None, step=None): return cls(range_of(a, b=b, step=step))
    225 
--> 226     def map(self, f, *args, gen=False, **kwargs): return self._new(map_ex(self, f, *args, gen=gen, **kwargs))
    227     def argwhere(self, f, negate=False, **kwargs): return self._new(argwhere(self, f, negate, **kwargs))
    228     def filter(self, f=noop, negate=False, gen=False, **kwargs):

/opt/conda/lib/python3.7/site-packages/fastcore/basics.py in map_ex(iterable, f, gen, *args, **kwargs)
    541     res = map(g, iterable)
    542     if gen: return res
--> 543     return list(res)
    544 
    545 # Cell

/opt/conda/lib/python3.7/site-packages/fastcore/basics.py in __call__(self, *args, **kwargs)
    531             if isinstance(v,_Arg): kwargs[k] = args.pop(v.i)
    532         fargs = [args[x.i] if isinstance(x, _Arg) else x for x in self.pargs] + args[self.maxi+1:]
--> 533         return self.func(*fargs, **kwargs)
    534 
    535 # Cell

/opt/conda/lib/python3.7/site-packages/fastai/learner.py in _call_one(self, event_name)
    135     def _call_one(self, event_name):
    136         assert hasattr(event, event_name), event_name
--> 137         [cb(event_name) for cb in sort_by_run(self.cbs)]
    138 
    139     def _bn_bias_state(self, with_bias): return norm_bias_params(self.model, with_bias).map(self.opt.state)

/opt/conda/lib/python3.7/site-packages/fastai/learner.py in <listcomp>(.0)
    135     def _call_one(self, event_name):
    136         assert hasattr(event, event_name), event_name
--> 137         [cb(event_name) for cb in sort_by_run(self.cbs)]
    138 
    139     def _bn_bias_state(self, with_bias): return norm_bias_params(self.model, with_bias).map(self.opt.state)

/opt/conda/lib/python3.7/site-packages/fastai/callback/core.py in __call__(self, event_name)
     42                (self.run_valid and not getattr(self, 'training', False)))
     43         res = None
---> 44         if self.run and _run: res = getattr(self, event_name, noop)()
     45         if event_name=='after_fit': self.run=True #Reset self.run to True at each end of fit
     46         return res

<ipython-input-84-f5d8be49ab6b> in before_batch(self)
     21         if not self.stack_y:
     22             ny_dims = len(self.y.size())
---> 23             self.learn.yb = tuple(L(self.yb1,self.yb).map_zip(torch.lerp,weight=unsqueeze(self.lam, n=ny_dims-1)))

/opt/conda/lib/python3.7/site-packages/fastcore/foundation.py in map_zip(self, f, cycled, *args, **kwargs)
    248     def zip(self, cycled=False): return self._new((zip_cycle if cycled else zip)(*self))
    249     def zipwith(self, *rest, cycled=False): return self._new([self, *rest]).zip(cycled=cycled)
--> 250     def map_zip(self, f, *args, cycled=False, **kwargs): return self.zip(cycled=cycled).starmap(f, *args, **kwargs)
    251     def map_zipwith(self, f, *rest, cycled=False, **kwargs):
    252         return self.zipwith(*rest, cycled=cycled).starmap(f, **kwargs)

/opt/conda/lib/python3.7/site-packages/fastcore/foundation.py in starmap(self, f, *args, **kwargs)
    245         return first(self.map(f, *args, gen=False, **kwargs), g)
    246 
--> 247     def starmap(self, f, *args, **kwargs): return self._new(itertools.starmap(partial(f,*args,**kwargs), self))
    248     def zip(self, cycled=False): return self._new((zip_cycle if cycled else zip)(*self))
    249     def zipwith(self, *rest, cycled=False): return self._new([self, *rest]).zip(cycled=cycled)

/opt/conda/lib/python3.7/site-packages/fastcore/foundation.py in _new(self, items, *args, **kwargs)
    180     @property
    181     def _xtra(self): return None
--> 182     def _new(self, items, *args, **kwargs): return type(self)(items, *args, use_list=None, **kwargs)
    183     def __getitem__(self, idx): return self._get(idx) if is_indexer(idx) else L(self._get(idx), use_list=None)
    184     def copy(self): return self._new(self.items.copy())

/opt/conda/lib/python3.7/site-packages/fastcore/foundation.py in __call__(cls, x, *args, **kwargs)
    161     def __call__(cls, x=None, *args, **kwargs):
    162         if not args and not kwargs and x is not None and isinstance(x,cls): return x
--> 163         return super().__call__(x, *args, **kwargs)
    164 
    165 # Cell

/opt/conda/lib/python3.7/site-packages/fastcore/foundation.py in __init__(self, items, use_list, match, *rest)
    171         if items is None: items = []
    172         if (use_list is not None) or not is_array(items):
--> 173             items = list(items) if use_list else listify(items)
    174         if match is not None:
    175             if is_coll(match): match = len(match)

/opt/conda/lib/python3.7/site-packages/fastcore/basics.py in listify(o)
     49     if isinstance(o, list): return o
     50     if isinstance(o, str) or is_array(o): return [o]
---> 51     if is_iter(o): return list(o)
     52     return [o]
```

Place between these lines with triple backticks:

```
```

**Additional context**

**Comment by jaideep11061982** [comment]
This is the line that causes the error

`  self.learn.yb = tuple(L(self.yb1,self.yb).map_zip(torch.lerp,weight=unsqueeze(self.lam, n=ny_dims-1)))`

**Comment by jaideep11061982** [comment]
close this issue. I figured it out. y_int=True was missing

**Comment by hritiknarayan** [comment]
@jaideep11061982 could you tell me what exact changes you made to resolve this bug?
